### PR TITLE
allow lowercase postcodes

### PR DESCRIPTION
**Pull request by RossKen**
Fix regex to allow lowercase postcodes

### CODE REVIEW

File: splink/comparison_template_library.py [hunk 1]
```diff
@@ -1036,7 +1036,7 @@ def __init__(
         self,
         col_name: str,
         invalid_postcodes_as_null=False,
-        valid_postcode_regex="^[A-Z]{1,2}[0-9][A-Z0-9]? [0-9][A-Z]{2}$",
+        valid_postcode_regex="^[A-Za-z]{1,2}[0-9][A-Za-z0-9]? [0-9][A-Za-z]{2}$",
```
Comment (ThomasHepworth) [inline splink/comparison_template_library.py:1039]: Should we add an argument that fixes the casing for your column?

Comment (RossKen) [reply splink/comparison_template_library.py:1039]: Hmm, potentially

I have just realised that the current changes won't fix the exact match level

Comment (RossKen) [reply splink/comparison_template_library.py:1039]: Which would get fixed if we cast the column

Comment (ThomasHepworth) [reply splink/comparison_template_library.py:1040]: Is this snippet of code used?
```
postcode_col = InputColumn(col_name, sql_dialect=self._sql_dialect)
postcode_col_l, postcode_col_r = postcode_col.names_l_r()
```

Comment (RossKen) [reply splink/comparison_template_library.py:1040]: No, but I have removed in the string comparators pr I'm about to put the pngs in so was going to leave it in this one

File: splink/comparison_template_library.py [hunk 2]
```diff
@@ -1170,6 +1174,9 @@ def __init__(
 
         comparison_levels = []
 
+        if set_to_uppercase:
+            postcode_col = f"upper({col_name})"
```
Comment (ThomasHepworth) [inline splink/comparison_template_library.py:1178]: I think this should be using `postcode_col.sql()` as the column name.

If we encounter an invalid column name (a name with spaces, a name such as "group", etc), this code breaks, I believe.

Comment (ThomasHepworth) [reply splink/comparison_template_library.py:1178]: Yeah, just to confirm, this doesn't appear to work.

You'd need to put `postcode_col` inside all of the existing `_exact_match_level` calls.

Comment (RossKen) [reply splink/comparison_template_library.py]: Sorry, I think this was wrong. Have changed to `col_name = f"upper({col_name})"` to ensure the capitalisation pulls through the rest of the comparison levels

Comment (ThomasHepworth) [reply splink/comparison_template_library.py]: That also works.

**Review by ThomasHepworth — APPROVED** [review]

**Comment by RossKen** [comment]
@sama-ds and I have found a similar issue with capital letters in emails. Given this is a general problem - I am going to add a `set_to_lowercase` parameter to the comparison_level_library_functions that can be pulled through for different `cl` and `ctl` functions.

Also, doing some digging, capitalisation is potentially leading to missing a number of TPs (if users are not standardising their data properly) as comparator functions are case sensitive

image (5)

File: splink/comparison_level_library.py [hunk 3]
```diff
@@ -86,6 +86,7 @@ def __init__(
         self,
         col_name,
         regex_extract: str = None,
+        set_to_lowercase: bool = False,
```
Comment (ThomasHepworth) [inline splink/comparison_level_library.py:89]: Do you know how this functions when we aren't comparing strings?

Comment (RossKen) [reply splink/comparison_level_library.py:89]: Good point - I will check this afternoon

Comment (RossKen) [reply splink/comparison_level_library.py:89]: In duckdb it runs without error - will check in spark now

Comment (RossKen) [reply splink/comparison_level_library.py:89]: Yea seems to work fine for dates and numerics for both spark and duckdb

Comment (ThomasHepworth) [reply splink/comparison_level_library.py:89]: I'm happy for you to merge this in that case.

**Review by ThomasHepworth — APPROVED** [review]

**Pull request merged**